core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -487,7 +623,77 @@ mkScottConstructorBody resultTypeName caseNamesAndTypes argNamesAndTypes index m
         afuncs = mkIterLamAbs argNamesAndTypes fixed
     in afuncs
 
-convAlt :: Converting m => Bool -> GHC.DataCon -> GHC.CoreAlt -> m PLCExpr
+getConstructors :: Converting m => GHC.TyCon -> m [PLCTerm]
+getConstructors tc = do
+    -- make sure the constructors have been created
+    _ <- convTyCon tc
+    defs <- gets csTypeDefs
+    case Map.lookup (GHC.getName tc) defs of
+        Just (tydConstrs -> Just constrs, _) -> pure $ tdTerm <$> constrs
+        _                                    -> throwPlain $ ConversionError "Constructors have not been converted"
+
+getConstructorsInstantiated :: Converting m => GHC.Type -> m [PLCTerm]
+getConstructorsInstantiated = \case
+    (GHC.splitTyConApp_maybe -> Just (tc, args)) -> do
+        constrs <- getConstructors tc
+
+        forM constrs $ \c -> do
+            args' <- mapM convType args
+            pure $ mkIterInst c args'
+    -- must be a TC app
+    _ -> throwPlain $ ConversionError "Type was not a type constructor application"
+
+mkMatchTy :: Converting m => GHC.TyCon -> [GHC.DataCon] -> m PLCType
+mkMatchTy tc dcs =
+    let
+        tyVars = GHC.tyConTyVars tc
+    in
+        flip (foldr (\tv acc -> mkTyForall tv acc)) tyVars $ do
+            tc' <- convTyCon tc
+            args <- mapM (convType . GHC.mkTyVarTy) tyVars
+            let applied = mkIterTyApp tc' args
+
+            scottTy <- mkScottTy tc dcs
+
+            pure $ PLC.TyFun () applied scottTy
+
+mkMatch :: Converting m => GHC.TyCon -> m PLCTerm
+mkMatch tc =
+    let
+        tyVars = GHC.tyConTyVars tc
+    in
+        flip (foldr (\tv acc -> mkTyAbs tv acc)) tyVars $ do
+            tc' <- convTyCon tc
+            args <- mapM (convType . GHC.mkTyVarTy) tyVars
+            let applied = mkIterTyApp tc' args
+
+            x <- safeFreshName "x"
+            isSelfRecursive <- isSelfRecursiveTyCon tc
+            let body = PLC.Var () x
+            let unwrapped = if isSelfRecursive then PLC.Unwrap () body else body
+
+            pure $ PLC.LamAbs () x applied unwrapped
+
+getMatch :: Converting m => GHC.TyCon -> m PLCTerm
+getMatch tc = do
+    -- ensure the tycon has been converted, which will create the matcher
+    _ <- convTyCon tc
+    defs <- gets csTypeDefs
+    case Map.lookup (GHC.getName tc) defs of
+        Just (tydMatch -> Just match, _) -> pure $ tdTerm match
+        _                                -> throwPlain $ ConversionError "Match has not been converted"
+
+getMatchInstantiated :: Converting m => GHC.Type -> m PLCTerm
+getMatchInstantiated = \case
+    (GHC.splitTyConApp_maybe -> Just (tc, args)) -> do
+        match <- getMatch tc
+
+        args' <- mapM convType args
+        pure $ mkIterInst match args'
+    -- must be a TC app
+    _ -> throwPlain $ ConversionError "Type was not a type constructor application"
+
+convAlt :: Converting m => Bool -> GHC.DataCon -> GHC.CoreAlt -> m PLCTerm
```

Comment:
This function and several above are not commented. Really hard to review.

Reply:
Yes, I should document these more.